### PR TITLE
Play next N messages

### PR DESCRIPTION
**Pull request by agalbachicar**
- Modifies the Player interface to accept an optional parameter stating the number of messages to play.
- Modifies the the PlayNext.srv message to receive the number of messages to play. Default is 1.
- Tests were included to verify affected code.

### CODE REVIEW

File: rosbag2_transport/src/rosbag2_transport/player.cpp [hunk 1]
```diff
@@ -357,27 +359,44 @@ rosbag2_storage::SerializedBagMessageSharedPtr * Player::peek_next_message_from_
   return message_ptr;
 }
 
-bool Player::play_next()
+bool Player::play_next(std::optional<uint64_t> num_messages)
```
Comment (agalbachicar) [inline rosbag2_transport/src/rosbag2_transport/player.cpp]:
```suggestion
bool Player::play_next(const std::optional<uint64_t> num_messages)
```

**Comment by agalbachicar** [comment]
@gbiggs for review.

File: rosbag2_transport/src/rosbag2_transport/player.cpp [hunk 2]
```diff
@@ -357,27 +359,44 @@ rosbag2_storage::SerializedBagMessageSharedPtr * Player::peek_next_message_from_
...
+      message_queue_.pop();
+      message_ptr = peek_next_message_from_queue();
+    }
+    // There are no more messages of the built publishers to play.
+    if (!next_message_published) {
+      break;
     }
-    message_queue_.pop();
-    message_ptr = peek_next_message_from_queue();
+    message_countdown--;
```
Comment (gbiggs) [inline rosbag2_transport/src/rosbag2_transport/player.cpp:401]: Should this be inside the inner `while` loop?

File: rosbag2_transport/test/rosbag2_transport/test_play_next.cpp [hunk 3]
```diff
@@ -357,3 +420,47 @@ TEST_F(RosBag2PlayTestFixture, play_next_playing_only_filtered_topics) {
   // All we care is that any messages arrived
   EXPECT_THAT(replayed_topic2, SizeIs(Eq(3u)));
 }
+
+TEST_F(RosBag2PlayTestFixture, play_next_no_message_must_success) {
```
Comment (gbiggs) [inline rosbag2_transport/test/rosbag2_transport/test_play_next.cpp]:
```suggestion
TEST_F(RosBag2PlayTestFixture, play_next_no_message_must_succeed) {
```

File: rosbag2_transport/src/rosbag2_transport/player.cpp (hunk 2 above)
Comment (agalbachicar) [reply rosbag2_transport/src/rosbag2_transport/player.cpp:401]: I've added a docstring in 0381e31 . `message_countdown` should be decreased when a message is effectively published. `publish_message()` already cares of sending the message iff there is matching publisher for its topic. When there is none, the return value is `false` and the previous while loop will continue if there are more messages in the storage. On the contrary, when the message is sent, the flag becomes true and the while loop condition is no longer valid which also makes the next if clause to be skipped and the counter to be decreased.

I hope that docstring helps with the understanding of the code without needing to go back and forth in the file (I had to do it when inspecting similar loops in other functions).

**Comment by agalbachicar** [comment]
It's ready for another review @gbiggs . Thanks!

**Review by gbiggs — APPROVED** [review]

**Pull request merged**